### ISSUE TITLE
[Feature Request] Add ZodObject support to .pick & .omit

### ISSUE THREAD

**Issue by jfsiii** [colinhacks/zod#285]
As I understand it, `.pick` & `.omit` currently accept `Record<string, true | undefined>`. I think it would be useful if it accepted a zod value as well.

For example, if we start with 

```javascript
const A = z.object({ a: z.number(), b: z.string() });
const B = z.object({ c: z.string(), d: z.string() });
const AB = A.merge(B);
```

and wanted to get the `A` schema back from `AB` we need to do

```javascript
const A2 = AB.pick({ a: true, b: true });
```

 or something like this if we wanted to ensure we always got all properties from `A`
```javascript
const aKeys = Object.keys(A.shape).reduce((acc, key) => {
  acc[key] = true;
  return acc;
}, {} as Record<string, true>);
const A3 = AB.pick(aKeys);
```

Is there a reason the values must be `true`? Is "truthy" enough? That would allow `AB.pick(A.shape)`

I think `string[]` array is closest to TS `Pick`,
```typescript
type A = Pick<A, 'a' | 'b'>
const A = AB.pick(Object.keys(A.shape));
```

but that would be a breaking change. Is there an issue or any other docs/background about using object vs array as the second arg?

I only found `zod` yesterday and spent a little time reading docs and trying some examples, so I could be missing/misunderstanding something. I just wondered if we needed this new interface or if we could use an existing one.

Cheers on the great library.

**Comment by colinhacks** [comment]
You don't seem to be misunderstanding anything! 

The hard part is making everything work in the *static* domain, not the runtime domain. If ZodObject.pick accepted an input of type `string[]` the TypeScript engine wouldn't have any way of knowing which keys exist or don't exist on the resulting object (post-`pick`). 

The reason the input to pick is an object is because TypeScript is good at infering the keys of objects. I recommend playing around with this function to get a sense of what TypeScript can infer:

```ts
const infer = <T>(arg: T): T => {
  return arg;
};
```

Allowing any truthy value is a bit messy, it would require re-implementing JavaScript's weird truthiness rules in TypeScript. It's also hard for other reasons, like this:

```ts
AB.pick({
  key1: 'keep',
  key2: ''
})
```

TypeScript infers this as `{ key1:string, key2: string}` (no difference between the two values) but one would get picked and the other omitted. So I wouldn't be able to return a correct static type. 

TLDR: Inference is hard and messy, there's (usually) a reason Zod is implemented like it is.

**Comment by davidmdm** [comment]
I can help you out here if you want to change your api for zod3. 

myzod implements pick and omit in the way described here.
Here is the version light of how it is done:

```typescript
class ObjectType<T extends Record<string, AnyType>> {

   ... 

   pick<K extends StringTypes<keyof T>>(
        keys: K[]
     ): ObjectType<Pick<T, Extract<StringTypes<keyof T>, ToUnion<typeof keys>>>>

}
```

where:
```typescript
type StringTypes<T> = T extends string ? T : never;
type ToUnion<T extends any[]> = T[number];
```
Hope that helps if you ever want to go down this road!

**Comment by ntsim** [comment]
I was trying pick/omit today and was kinda hoping to be able to toggle properties based on the truthiness, but was surprised to see that just having the property will end up omiting/picking it, even if it's set to `undefined`.

Even if a key-based API isn't used, it would be good if it could accept actual true/false values for property toggling. Seems like it should be possible to do so with something like this:

```typescript
type KeysWithType<T, U> = {
  [K in keyof T]: T[K] extends U ? K : never;
}[keyof T];

type PickByType<T, U> = Pick<T, KeysWithType<T, U>>;

interface ObjectType<T> {
  omit<Mask extends { [Key in keyof T]?: boolean }>(
    mask: Mask,
  ): Omit<T, keyof PickByType<Mask, true>>;
}
```
